python/tests/test_tree_stats.py

```diff
@@ -3604,23 +3759,29 @@ def branch_allele_frequency_spectrum(
     branch_length = np.zeros(ts.num_nodes)
     tree_index = 0
 
-    def update_result(window_index, u, right):
-        if 0 < count[u, -1] < ts.num_samples:
-            x = (right - last_update[u]) * branch_length[u]
-            c = count[u, :num_sample_sets]
-            if not polarised:
-                c = fold(c, out_dim)
-            index = tuple([window_index] + list(c))
-            result[index] += x
+    def update_result(window_index, u, right, time_windows):
+        for k_tw, _ in enumerate(time_windows[:-1]):
+            if 0 < count[u, -1] < ts.num_samples:
+                # interval between child and parent inside the window
+                t_v = branch_length[u] + time[u]
```

Comment:
Here we're losing the advantage of cacheing `branch_length[u]`; this might as well be `time[v]` (with `v` passed in also).

Alternatives:
1. let `branch_length` be a (num nodes x num time windows) array instead of just a vector, so that we'd have
```
tw_branch_length = branch_length[u, k_tw]
```
2. Do the calculation [down a few lines](https://github.com/tskit-dev/tskit/pull/2948/files#diff-4d6ee7e1fdeb228e5a8a5fe5c37af121dc87237d0c30563eafab80282fe3ed0eR3781), something like this (this is not right):
```
u = edge.child
v = edge.parent
t_c = time[u]
t_p = time[v]
time_window_index = 0
while t_p < time_windows[time_window_index + 1]:
    while v != -1:
        tw_branch_length = min(time_windows[k_tw + 1], t_p) - max(time_windows[k_tw], t_c)
        update_result(window_index, time_window_index, v, t_left, tw_branch_length)
        count[v] -= count[u]
        t_c = t_p
        v = parent[v]
        t_p = time[v]
    time_window_index += 1
```
The advantage to this is that computation isn't increased by a factor of (num time windows). The disadvantage might be that the code is harder to understand?